### PR TITLE
[Dynamo] Support dict.setdefault

### PR DESCRIPTION
**Pull request by yanboliang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134083

cc @rec

### CODE REVIEW

File: test/dynamo/test_functions.py [hunk 1]
```diff
@@ -587,6 +587,33 @@ def test_dict_values(x):
...
+    def test_dict_setdefault2(x):
+        d = {"a": 1, "b": 2}
+        d.setdefault("c", 10)
+        if d["c"] == 10:
+            return x + 1
+        else:
+            return x - 1
+
+    @make_test
+    def test_dict_setdefault3(x):
```
Comment (XuehaiPan) [inline test/dynamo/test_functions.py:609]: We may also need to test `defaultdict.setdefault`.

File: torch/_dynamo/variables/dicts.py [hunk 2]
```diff
@@ -335,6 +335,19 @@ def call_method(
             return self.getitem_const(tx, args[0])
         elif name == "__contains__" and len(args) == 1:
             return ConstantVariable.create(args[0] in self)
+        elif name == "setdefault" and arg_hashable and self.mutable_local:
+            assert not kwargs and len(args) <= 2
```
Comment (Skylion007) [inline torch/_dynamo/variables/dicts.py:339]: split into two asserts so it's easier to know which condition failed:
```suggestion
            assert not kwargs
            assert len(args) <= 2
```

Comment (williamwen42) [reply torch/_dynamo/variables/dicts.py:339]: May be better to use `unimplemented`.

Comment (yanboliang) [reply torch/_dynamo/variables/dicts.py]: It's illegal argument, so it should be assertion.

Comment (yanboliang) [reply torch/_dynamo/variables/dicts.py]: chatted with @williamwen42 offline, unimplemented seems a more correct way, though we use assert in many other places.

**Review by williamwen42 — APPROVED** [review]

**Pull request closed without merging**